addons/crm/models/crm_lead.py

```diff
@@ -749,6 +749,7 @@ def action_set_automated_probability(self):
     def action_set_won_rainbowman(self):
         self.ensure_one()
         self.action_set_won()
+        self.flush(['stage_id'])
```

Comment:
Assuming this is because `_get_rainbowman_message` accesses things in SQL, wouldn't it make more sense for this to be in `_get_rainbowman_message` (to ensure the db is in the correct state)? Possibly alongside the other fields influencing `probability` just in case?

Reply:
Also might be a good idea to ask the ORM for all the dependencies of `probability` and flush them all, rather than assume only `stage_id` could be an issue?

Reply:
Actually, this flush call was making things work but I'm not sure how, must have been a side effect.

What I should do instead is flushing probability in the `_get_rainbowman_message`, since it's the stored field that is modified just before calling this method (in `action_set_won`).
Other fields that are used within the raw SQL could be flushed but none seem problematic from what I have tested and I would rather only flush what is absolutely necessary.

I changed it and it seems to work fine and make more sense.
Feel free to tell me if you think it's still not a good idea.

Thanks for the review!

Reply:
> Actually, this flush call was making things work but I'm not sure how, must have been a side effect.

I think it's because `stage_id` is a dependency of `probability` which the SQL request uses. So when `action_set_won` updates the `stage_id`, `probability` doesn't get updated in-database until it's accessed or `stage_id` is flushed.

Reply:
Well `action_set_won` directly writes on probability:
`leads.write({'stage_id': won_stage_id.id, 'probability': 100})`

And I can confirm from my tests that without flushing probability, it doesn't work, and it works when I flush it.
So it's doing "something right", but it may not be the best / most logical option?

Reply:
> Well action_set_won directly writes on probability:

Oh I'd missed that bit, so yeah it works under the assumption that probability is always straight written up. But we may also want to flush the fields on which probability depends in case it's one of those which triggers the change, that seems more correct.

Or maybe just flush everything to be safe, if this is not too common a call / demanding?

Reply:
Ok let's just flush everything then.
This method is only called when "needed" (meaning when a lead enters a "won" stage) so it should be fine in terms of performances.

The "probability" field is a wild beast and computing the fields it depends on might be as much overhead as flushing everything anyway.